miniwin/miniwin/src/miniwin_d3drmmesh.cpp

```diff
@@ -25,7 +33,23 @@ HRESULT Direct3DRMMeshImpl::AddGroup(
 	D3DRMGROUPINDEX* groupIndex
 )
 {
-	MINIWIN_NOT_IMPLEMENTED();
+	if (faceCount < 0 || vertexPerFace < 0 || !faceBuffer) {
+		return DDERR_INVALIDPARAMS;
+	}
+
+	int newIndex = static_cast<int>(m_groups.size());
+	if (groupIndex) {
+		*groupIndex = newIndex;
+	}
```

Comment:
`IDirect3DRMMesh` only allows addition of groups. Can we use a `std::vector` for `m_groups` instead?
(Let's delete the copy constructor of `MeshGroup`, and implement the move constructor)

A `std::vector` also makes the bounds checks in the various getters a lot easier.

Reply:
Done by the reference counter is making me very confused on the material and texture, you might want to pull it in on your side and see if it's right.